### ISSUE TITLE
torch.mean(cpu) is implemented as torch.sum(x)/x.numel() which overflows on float16

### ISSUE THREAD

**Issue by lezcano** [pytorch/pytorch#91597]
### 🐛 Describe the bug

```python
>>> torch.ones(100000, dtype=torch.float16).mean()
tensor(nan, dtype=torch.float16)
```

This was documented in the implementation
https://github.com/pytorch/pytorch/blob/1a98c3e36cce0571f3915ade75d6f13225695ec4/aten/src/ATen/native/ReduceOps.cpp#L1314-L1317
and bit us here https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/8#discussion_r1060162705

### Versions

master

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by vadimkantorov** [comment]
Another example of reduction where it might be good to have acctype_t not equal to input dtype: https://github.com/pytorch/pytorch/issues/55366

**Comment by jgong5** [comment]
I would suggest we mark it as an enhancement rather than a bug since fp16 data type on CPU is not yet well supported or validated yet.

**Comment by lezcano** [comment]
I think it's a bug, because the results differ from those in CUDA, and it may give silently incorrect results. If `torch.float16` is not well supported on CPU we should warn the user, as we do with other prototype features.

**Comment by jgong5** [comment]
> I think it's a bug, because the results differ from those in CUDA, and it may give silently incorrect results. If `torch.float16` is not well supported on CPU we should warn the user, as we do with other prototype features.

Anyway, we will share the plan of adding complete CPU float16 support to PyTorch. @mingfeima is working on it.

**Comment by peterbell10** [comment]
`torch.sum` *is* accumulated in `float32` precision here, but the result 100000 is not representable as a `float16`. This is also an issue for `mean` because it is implemented with sum on CPU 
https://github.com/pytorch/pytorch/blob/1a98c3e36cce0571f3915ade75d6f13225695ec4/aten/src/ATen/native/ReduceOps.cpp#L1329

**Comment by lezcano** [comment]
Right, I updated the issue.
Note that this can be worked around for now by setting the dtype in this case to be `float16` if it's not specified and then casting the result back to the correct output dtype as discussed here https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/8#discussion_r1060565896

**Comment by mingfeima** [comment]
Oops, this one is more difficult than i thought...

The current situation is
* `mean_out` from ReduceOps.cpp has by passed original implementation with `mean_kernel_impl()`, see [ReduceOps.cpp#L1303](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L1303) due to the low performance.
* `mean_out` has a fast path with `at::sum() / numel`

The SumKernel from ATen will do sum with acc type, this holds true for both `bfloat16` and `float16`. The only difference is that `bfloat16` data loading is vectorized but `float16` data loading is not. I have it fixed with https://github.com/pytorch/pytorch/pull/96082

From the case in this thread, `sum()` will be `inf` and `numel` will also be `inf`, that's why we are getting `nan` on the cpu side.

The issue can be fixed in a similar manner to GPU side, take 2 steps to get it done :
* add a allreduce kernel for mean in `cpu/ReduceAllOpsKernel.cpp`.
* let `vec::reduce_all<Half>` return the final result in the acc type, aka, float. For the reduceall case, i don't think it is meaningful to convert the final result back to bfloat16 or float16.

Ideally we should try to fix `mean_kernel_impl()` performance issue, but this is going to be much harder, it will need at least 2 different optimization strategies for
* dim reduce on lastdim (takes 1 parallel path)
* dim reduce on non-lastdim (takes 2 parallel paths)

**Comment by mingfeima** [comment]
Fixed on nightly with https://github.com/pytorch/pytorch/pull/97351

current result:
```
>>> torch.ones(100000, dtype=torch.float16).mean()
tensor(1., dtype=torch.float16)
```
